### ISSUE TITLE
istio 0.8.0 didn't automatic injection sidecar

### ISSUE THREAD

**Issue by shenjieyao** [istio/istio#6504]
**Describe the bug**
I used a method of automatic injection sidecar into my service ,but didn't watch any istio-proxy in my service pod ! 
In Version 0.7.1 ,we can use following command to install webhook to support automatic injection 
"./install/kubernetes/webhook-create-signed-cert.sh --service istio-sidecar-injector --namespace istio-system --secret sidecar-injector-certs"
"kubectl apply -f  install/kubernetes/istio-sidecar-injector-configmap-release.yaml"
"cat install/kubernetes/istio-sidecar-injector.yaml | ./install/kubernetes/webhook-patch-ca-bundle.sh > install/kubernetes/istio-sidecar-injector-with-ca-bundle.yaml"
"kubectl apply -f install/kubernetes/istio-sidecar-injector-with-ca-bundle.yaml"
But in Version 0.8.0 , i can't found any steps to install webhook to support automatic injection . i think these steps weren't needed in new version ! so i felt confused to these parts when i did practice !

**Expected behavior**

**Steps to reproduce the bug**
1.“Kubectl get pods -n istio-system” and “Kubectl get pods -n istio-system” install control plane was ok 
2. “kubectl apply -f samples/sleep/sleep.yaml” to run sleep was ok ,and haven't istio-proxy was right.
3."kubectl label namespace default istio-injection=enabled" to add label was ok.
4."kubectl delete pod my_sleep_pod_name" ,my pod was terminated and haven't new pod was created .
5. "kubectl apply -f samples/sleep/sleep.yaml" ,i didn't watch any pod ! 
I tried to use manual injection was ok . but automatic injection was error ! i did not know whether i missed some steps ? 

**Version**
kubernetes version of client : 1.9.2
kubernetes version of service :1.10.0
Istio version :0.8.0
Minikube :0.28.0

**Is Istio Auth enabled or not?**
istio-demo-auth.yaml and istio-demo.yaml too

**Environment**
Linux version 4.4.0-116-generic (buildd@lgw01-amd64-021) (gcc version 5.4.0 20160609 (Ubuntu 5.4.0-6ubuntu1~16.04.9) ) #140-Ubuntu SMP Mon Feb 12 21
:23:04 UTC 2018

**Comment by ostromart** [comment]
Need some further information since this is a fairly basic scenario that has been well tested. It may be related to the environment or a missed step in the setup. 
Could you please include the following information:
 - what cluster type and version is this running on? e.g. minikube, gke 1.9.6 etc.
 - have you tried to reproduce the issue? does it happen every time?

**Comment by ymesika** [comment]
@shenjieyao Your steps are correct and the pod should be injected with a sidecar.

The step that you deleted the pod and K8s didn't start a new one instead implies that there is something wrong with the scheduling itself.
Take a look at the deployment to see if you get more hints: `kubectl describe deployment sleep`

**Comment by shenjieyao** [comment]
@ymesika  Yes ! I think so that it woring with the scheduling itself .but i don't know why i recreate sleep.yaml will be not watched any pod . i used that describe command ,and got following message !

Name:                   sleep
Namespace:              default
CreationTimestamp:      Fri, 22 Jun 2018 09:17:03 +0200
Labels:                 app=sleep
Annotations:            deployment.kubernetes.io/revision=1
                        kubectl.kubernetes.io/last-applied-configuration={"apiVersion":"extensions/v1beta1","kind":"Deployment","metadata":{"annotations":{},"name":"sleep","namespace":"default"},"spec":{"replicas":1,"templat...
Selector:               app=sleep
Replicas:               1 desired | 0 updated | 0 total | 0 available | 1 unavailable
StrategyType:           RollingUpdate
MinReadySeconds:        0
RollingUpdateStrategy:  1 max unavailable, 1 max surge
Pod Template:
  Labels:  app=sleep
  Containers:
   sleep:
    Image:      tutum/curl
    Port:       <none>
    Host Port:  <none>
    Command:
      /bin/sleep
      infinity
    Environment:  <none>
    Mounts:       <none>
  Volumes:        <none>
Conditions:
  Type             Status  Reason
  ----             ------  ------
  Available        True    MinimumReplicasAvailable
  Progressing      True    NewReplicaSetAvailable
  ReplicaFailure   True    FailedCreate
OldReplicaSets:    <none>
NewReplicaSet:     sleep-86f6b99f94 (0/1 replicas created)
Events:
  Type    Reason             Age   From                   Message
  ----    ------             ----  ----                   -------
  Normal  ScalingReplicaSet  2m    deployment-controller  Scaled up replica set sleep-86f6b99f94 to 1  

I think it was error ,right ?

**Comment by ymesika** [comment]
Yes, it failed to recreate an instance.
Get the exact reason for the failure with `kubectl describe replicaset xxxxx`.

**Comment by shenjieyao** [comment]
Thanks , I got following error message ! but I didn't understand it . i'm a newcomer  ! ^^
Events:
  Type     Reason            Age                From                   Message
  ----     ------            ----               ----                   -------
  Normal   SuccessfulCreate  25m                replicaset-controller  Created pod: sleep-86f6b99f94-cr695
  Warning  FailedCreate      6m (x19 over 23m)  replicaset-controller  Error creating: Internal error occurred: failed calling admission webhook "sidecar-injector.istio.io": Post https://istio-sidecar-injector.istio-system.svc:443/inject?timeout=30s: Not Found

**Comment by ymesika** [comment]
That's very helpful. Can you get all pods from istio-system? It seems like the injector isn't up

**Comment by shenjieyao** [comment]
Yes ! all of them were up !
NAME                                       READY     STATUS      RESTARTS   AGE
grafana-6f6dff9986-snn9x                   1/1       Running     0          2h
istio-citadel-7bdc7775c7-vdf7n             1/1       Running     0          2h
istio-cleanup-old-ca-ct4bb                 0/1       Completed   0          2h
istio-egressgateway-795fc9b47-tdg6x        1/1       Running     0          2h
istio-ingressgateway-7d89dbf85f-ctcpp      1/1       Running     0          2h
istio-mixer-post-install-qwg2b             0/1       Completed   0          2h
istio-pilot-66f4dd866c-svwpv               2/2       Running     0          2h
istio-policy-76c8896799-fkxdt              2/2       Running     0          2h
istio-sidecar-injector-645c89bc64-nl2dk    1/1       Running     0          2h
istio-statsd-prom-bridge-949999c4c-knm6g   1/1       Running     0          2h
istio-telemetry-6554768879-4622z           2/2       Running     0          2h
istio-tracing-754cdfd695-gtm9d             1/1       Running     0          2h
prometheus-86cb6dd77c-9nnzt                1/1       Running     0          2h
servicegraph-5849b7d696-clprr              1/1       Running     0          2h

**Comment by ymesika** [comment]
Do you mind also getting the sidecar log?
`kubectl logs -n istio-system istio-sidecar-injector-645c89bc64-nl2dk`

**Comment by shenjieyao** [comment]
Following message is log istio-sidecar-injector  !

2018-06-22T05:34:32.550733Z	info	version root@48d5ddfd72da-docker.io/istio-0.8.0-6f9f420f0c7119ff4fa6a1966a6f6d89b1b4db84-Clean
2018-06-22T05:34:32.551867Z	info	New configuration: sha256sum 6fb54ce8cab658754b93e2064ea052a1cd0682a6a385ea03d4ebe66707327e2a
2018-06-22T05:34:32.551896Z	info	Policy: enabled
2018-06-22T05:34:32.551940Z	info	Template: |
  initContainers:
  - name: istio-init
    image: docker.io/istio/proxy_init:0.8.0
    args:
    - "-p"
    - [[ .MeshConfig.ProxyListenPort ]]
    - "-u"
    - 1337
    - "-m"
    - [[ or (index .ObjectMeta.Annotations "sidecar.istio.io/interceptionMode") .ProxyConfig.InterceptionMode.String ]]
    - "-i"
    [[ if (isset .ObjectMeta.Annotations "traffic.sidecar.istio.io/includeOutboundIPRanges") -]]
    - "[[ index .ObjectMeta.Annotations "traffic.sidecar.istio.io/includeOutboundIPRanges"  ]]"
    [[ else -]]
    - "*"
    [[ end -]]
    - "-x"
    [[ if (isset .ObjectMeta.Annotations "traffic.sidecar.istio.io/excludeOutboundIPRanges") -]]
    - "[[ index .ObjectMeta.Annotations "traffic.sidecar.istio.io/excludeOutboundIPRanges"  ]]"
    [[ else -]]
    - ""
    [[ end -]]
    - "-b"
    [[ if (isset .ObjectMeta.Annotations "traffic.sidecar.istio.io/includeInboundPorts") -]]
    - "[[ index .ObjectMeta.Annotations "traffic.sidecar.istio.io/includeInboundPorts"  ]]"
    [[ else -]]
    - [[ range .Spec.Containers -]][[ range .Ports -]][[ .ContainerPort -]], [[ end -]][[ end -]][[ end]]
    - "-d"
    [[ if (isset .ObjectMeta.Annotations "traffic.sidecar.istio.io/excludeInboundPorts") -]]
    - "[[ index .ObjectMeta.Annotations "traffic.sidecar.istio.io/excludeInboundPorts" ]]"
    [[ else -]]
    - ""
    [[ end -]]
    imagePullPolicy: IfNotPresent
    securityContext:
      capabilities:
        add:
        - NET_ADMIN
      privileged: true
    restartPolicy: Always
  
  containers:
  - name: istio-proxy
    image: [[ if (isset .ObjectMeta.Annotations "sidecar.istio.io/proxyImage") -]]
    "[[ index .ObjectMeta.Annotations "sidecar.istio.io/proxyImage" ]]"
    [[ else -]]
    docker.io/istio/proxyv2:0.8.0
    [[ end -]]
    args:
    - proxy
    - sidecar
    - --configPath
    - [[ .ProxyConfig.ConfigPath ]]
    - --binaryPath
    - [[ .ProxyConfig.BinaryPath ]]
    - --serviceCluster
    [[ if ne "" (index .ObjectMeta.Labels "app") -]]
    - [[ index .ObjectMeta.Labels "app" ]]
    [[ else -]]
    - "istio-proxy"
    [[ end -]]
    - --drainDuration
    - [[ formatDuration .ProxyConfig.DrainDuration ]]
    - --parentShutdownDuration
    - [[ formatDuration .ProxyConfig.ParentShutdownDuration ]]
    - --discoveryAddress
    - [[ .ProxyConfig.DiscoveryAddress ]]
    - --discoveryRefreshDelay
    - [[ formatDuration .ProxyConfig.DiscoveryRefreshDelay ]]
    - --zipkinAddress
    - [[ .ProxyConfig.ZipkinAddress ]]
    - --connectTimeout
    - [[ formatDuration .ProxyConfig.ConnectTimeout ]]
    - --statsdUdpAddress
    - [[ .ProxyConfig.StatsdUdpAddress ]]
    - --proxyAdminPort
    - [[ .ProxyConfig.ProxyAdminPort ]]
    - --controlPlaneAuthPolicy
    - [[ .ProxyConfig.ControlPlaneAuthPolicy ]]
    env:
    - name: POD_NAME
      valueFrom:
        fieldRef:
          fieldPath: metadata.name
    - name: POD_NAMESPACE
      valueFrom:
        fieldRef:
          fieldPath: metadata.namespace
    - name: INSTANCE_IP
      valueFrom:
        fieldRef:
          fieldPath: status.podIP
    - name: ISTIO_META_POD_NAME
      valueFrom:
        fieldRef:
          fieldPath: metadata.name
    - name: ISTIO_META_INTERCEPTION_MODE
      value: [[ or (index .ObjectMeta.Annotations "sidecar.istio.io/interceptionMode") .ProxyConfig.InterceptionMode.String ]]
    imagePullPolicy: IfNotPresent
    securityContext:
        privileged: false
        readOnlyRootFilesystem: true
        [[ if eq (or (index .ObjectMeta.Annotations "sidecar.istio.io/interceptionMode") .ProxyConfig.InterceptionMode.String) "TPROXY" -]]
        capabilities:
          add:
          - NET_ADMIN
        [[ else -]]
        runAsUser: 1337
        [[ end -]]
    restartPolicy: Always
    resources:
      requests:
        cpu: 100m
        memory: 128Mi
      
    volumeMounts:
    - mountPath: /etc/istio/proxy
      name: istio-envoy
    - mountPath: /etc/certs/
      name: istio-certs
      readOnly: true
  volumes:
  - emptyDir:
      medium: Memory
    name: istio-envoy
  - name: istio-certs
    secret:
      optional: true
      [[ if eq .Spec.ServiceAccountName "" -]]
      secretName: istio.default
      [[ else -]]
      secretName: [[ printf "istio.%s" .Spec.ServiceAccountName ]]
      [[ end -]]
2018-06-22T05:34:32.552743Z	warn	Neither --kubeconfig nor --master was specified.  Using the inClusterConfig.  This might not work.

**Comment by sdake** [comment]
I was able to reproduce what appears like the Helm chart not rendering properly on 0.8.0 in the sidecar injector.

e.g.:
```
sdake@falkor-07:~/istio-0.8.0/install/kubernetes$ kubectl logs istio-sidecar-injector-645c89bc64-wslq5 -n istio-system
2018-06-23T16:36:51.420665Z	info	version root@48d5ddfd72da-docker.io/istio-0.8.0-6f9f420f0c7119ff4fa6a1966a6f6d89b1b4db84-Clean
2018-06-23T16:36:51.422732Z	info	New configuration: sha256sum 6fb54ce8cab658754b93e2064ea052a1cd0682a6a385ea03d4ebe66707327e2a
2018-06-23T16:36:51.422761Z	info	Policy: enabled
2018-06-23T16:36:51.422774Z	info	Template: |
  initContainers:
  - name: istio-init
    image: docker.io/istio/proxy_init:0.8.0
    args:
    - "-p"
    - [[ .MeshConfig.ProxyListenPort ]]
    - "-u"
    - 1337
    - "-m"
    - [[ or (index .ObjectMeta.Annotations "sidecar.istio.io/interceptionMode") .ProxyConfig.InterceptionMode.String ]]
    - "-i"
    [[ if (isset .ObjectMeta.Annotations "traffic.sidecar.istio.io/includeOutboundIPRanges") -]]
    - "[[ index .ObjectMeta.Annotations "traffic.sidecar.istio.io/includeOutboundIPRanges"  ]]"
    [[ else -]]
    - "*"
    [[ end -]]
    - "-x"
    [[ if (isset .ObjectMeta.Annotations "traffic.sidecar.istio.io/excludeOutboundIPRanges") -]]
    - "[[ index .ObjectMeta.Annotations "traffic.sidecar.istio.io/excludeOutboundIPRanges"  ]]"
    [[ else -]]
    - ""
    [[ end -]]
    - "-b"
    [[ if (isset .ObjectMeta.Annotations "traffic.sidecar.istio.io/includeInboundPorts") -]]
    - "[[ index .ObjectMeta.Annotations "traffic.sidecar.istio.io/includeInboundPorts"  ]]"
    [[ else -]]
    - [[ range .Spec.Containers -]][[ range .Ports -]][[ .ContainerPort -]], [[ end -]][[ end -]][[ end]]
    - "-d"
    [[ if (isset .ObjectMeta.Annotations "traffic.sidecar.istio.io/excludeInboundPorts") -]]
    - "[[ index .ObjectMeta.Annotations "traffic.sidecar.istio.io/excludeInboundPorts" ]]"
    [[ else -]]
    - ""
    [[ end -]]
...
```

**Comment by sdake** [comment]
Reading the git logs, `[[` and `]]` are delimiters for sidecar injection configuration.

**Comment by sdake** [comment]
I should have been more clear above.  I didn't duplicate the issue - just the helmish looking stuff from the sidecar configmap.  That turns out to not be a problem (I speculated the templates were rendering incorrectly and still somehow functional).

**Comment by ayj** [comment]
The injection configuration is also a template. Using helm we have a template of a template :/. We've used different delimiters to differentiate - `{{` and `}}` for the _outer_ helm template and `[[` and `]]` for the _inner_ injection template.

**Comment by shenjieyao** [comment]
Sorry ! I didn't clearly understand what your comments meaning  .Do I need to change sidecar configmap? or is this problem a bug and  you will fix it in milestone 1.1?

**Comment by ostromart** [comment]
A couple of fixes address the symptoms you describe - hopefully they will work for you:
https://github.com/istio/istio/pull/6610
https://github.com/istio/istio/pull/6435
Please update to the latest and see if the issue persists.

**Comment by shenjieyao** [comment]
I updated the latest version ,but still have this issue !

**Comment by ostromart** [comment]
Could you please confirm which version (git sha or release link) you used to try it?

**Comment by shenjieyao** [comment]
I used this command "curl -L https://git.io/getLatestIstio | sh -" to get latest version that is V0.8.0. I think that is this release "6f9f420f0c7119ff4fa6a1966a6f6d89b1b4db84" . And V0.7.1 release link is https://github.com/istio/istio/releases/download/0.7.1/istio-0.7.1-linux.tar.gz .

**Comment by sdake** [comment]
@ayj yup - I understood that after I made the statement in the bug log and indicated my misunderstanding in https://github.com/istio/istio/issues/6504#issuecomment-399700758 - although I guess that also wasn't clear.

@shenjieyao you may ignore my comments about the templating.  The implementation is normal and the results are correct.

**Comment by krancour** [comment]
I seem to be running into the same thing. I'm simply following documentation:

```
$ minikube start \
    --extra-config=controller-manager.cluster-signing-cert-file="/var/lib/localkube/certs/ca.crt" \
    --extra-config=controller-manager.cluster-signing-key-file="/var/lib/localkube/certs/ca.key" \
    --extra-config=apiserver.admission-control="NamespaceLifecycle,LimitRanger,ServiceAccount,PersistentVolumeLabel,DefaultStorageClass,DefaultTolerationSeconds,MutatingAdmissionWebhook,ValidatingAdmissionWebhook,ResourceQuota" \
    --kubernetes-version=v1.10.0
```

Modified this next command a bit to specify virtualization, resources, and k8s version, but the admission controllers are all straight out of the documentation:

```
minikube start \
	--vm-driver virtualbox \
	--cpus 2 \
	--memory 8192 \
	--disk-size 50g \
	--kubernetes-version=v1.10.4 \
	--extra-config=controller-manager.cluster-signing-cert-file="/var/lib/localkube/certs/ca.crt" \
	--extra-config=controller-manager.cluster-signing-key-file="/var/lib/localkube/certs/ca.key" \
	--extra-config=apiserver.admission-control="NamespaceLifecycle,LimitRanger,ServiceAccount,PersistentVolumeLabel,DefaultStorageClass,DefaultTolerationSeconds,MutatingAdmissionWebhook,ValidatingAdmissionWebhook,ResourceQuota"
```

Everything looks healthy:

```
istio-system   istio-citadel-85fcc767fc-7q7s5             1/1       Running     0          32m
istio-system   istio-egressgateway-867bc6b594-m4dr2       1/1       Running     0          32m
istio-system   istio-galley-6845cdfb8f-rhbx7              1/1       Running     0          32m
istio-system   istio-ingress-787bb7c94b-pvnkk             1/1       Running     0          32m
istio-system   istio-ingressgateway-6df8678998-lkgkw      1/1       Running     0          32m
istio-system   istio-mixer-post-install-dxtjz             0/1       Completed   0          31m
istio-system   istio-pilot-6db8d59464-4pkks               2/2       Running     0          32m
istio-system   istio-policy-7b978cf7df-hct67              2/2       Running     0          32m
istio-system   istio-sidecar-injector-855f88c954-lml9r    1/1       Running     0          32m
istio-system   istio-statsd-prom-bridge-59b45fd6d-bsdkb   1/1       Running     0          32m
istio-system   istio-telemetry-7dfcc797b6-d78zf           2/2       Running     0          32m
istio-system   prometheus-ffd95f9f6-srndm                 1/1       Running     0          32m
```

Install the bookinfo app:

```
kubectl apply -f samples/bookinfo/kube/bookinfo.yaml
```

Results don't look right. Note 1/1 containers ready for each pod-- which tells me none of those pods have Envoy sidecars:

```
default        details-v1-7b97668445-mxxfl                1/1       Running     0          27m
default        productpage-v1-7bbdd59459-ll4z4            1/1       Running     0          27m
default        ratings-v1-76dc7f6b9-m8rcb                 1/1       Running     0          27m
default        reviews-v1-64545d97b4-dtkb2                1/1       Running     0          27m
default        reviews-v2-8cb9489c6-6qq54                 1/1       Running     0          27m
default        reviews-v3-6bc884b456-zhvrc                1/1       Running     0          27m
```

For good measure:

```
$ minikube version
minikube version: v0.28.0

$ kubectl version
Client Version: version.Info{Major:"1", Minor:"10", GitVersion:"v1.10.4", GitCommit:"5ca598b4ba5abb89bb773071ce452e33fb66339d", GitTreeState:"clean", BuildDate:"2018-06-06T15:22:13Z", GoVersion:"go1.9.6", Compiler:"gc", Platform:"darwin/amd64"}
Server Version: version.Info{Major:"1", Minor:"10", GitVersion:"v1.10.4", GitCommit:"5ca598b4ba5abb89bb773071ce452e33fb66339d", GitTreeState:"clean", BuildDate:"2018-06-06T08:00:59Z", GoVersion:"go1.9.3", Compiler:"gc", Platform:"linux/amd64"}
```

**Comment by krancour** [comment]
I missed this:

```
$ kubectl label namespace default istio-injection=enablel
```

To be fair, although that does appear places in the documentation, it's quite absent from the [instructions for installing the bookinfo app](https://istio.io/docs/guides/bookinfo/) and that seems an ideal place to call attention this. I'll open a separate issue for that.

**Comment by shenjieyao** [comment]
Yes ! Absolutely you forgot to enable istio-injection label ,but i did !  After added this command and installed bookinfo app then those pods disappeared . I only watched deployment and service ! My minikube start configuration was different with yours , i used  "--vm-driver=none" .

**Comment by krancour** [comment]
> I'll open a separate issue for that.

https://github.com/istio/istio/issues/6957

**Comment by jansony1** [comment]
I came into the same error like shenjieyao with same configuration, except that I deploy my cluster on aws by KOPS~

**Comment by ostromart** [comment]
To summarize the state of this bug:
 - some documentation enhancement was appropriate for 0.8 (fixed https://github.com/istio/istio/issues/6957)
 - in 1.0, the injection selection criteria can be inverted to inject by default (if enableNamespacesByDefault is set) https://github.com/istio/istio/pull/6704/files. This is probably the behavior that most users want/expect, but is optional for now to facilitate transition
 - @jansony1 regarding the last comment, could you please open a new issue for that and fill in the template with info about your environment. If you're on 0.8 you may be hitting https://github.com/istio/istio/issues/6069 which has been fixed in 1.0.
